pallets/ddc-verification/src/lib.rs

```diff
@@ -1470,6 +1626,115 @@ pub mod pallet {
 			)))
 		}
 
+		pub(crate) fn challenge_sub_aggregates_not_in_consensus(
+			cluster_id: &ClusterId,
+			era_id: DdcEra,
+			dac_nodes: &[(NodePubKey, StorageNodeParams)],
+			bucket_node_aggregates_not_in_consensus: Vec<BucketNodeAggregatesActivity>,
+		) -> Result<Vec<BucketNodeAggregatesActivity>, Vec<OCWError>> {
+			let mut bucket_aggregates_passed_challenges: Vec<BucketNodeAggregatesActivity> = vec![];
+			let number_of_identifiers = T::MAX_MERKLE_NODE_IDENTIFIER;
+
+			log::info!(
+				"🚀 Challenge process starts when bucket sub aggregates are not in consensus!"
+			);
+
+			for bucket_node_aggregate_activity in bucket_node_aggregates_not_in_consensus {
+				let merkle_node_ids = Self::find_random_merkle_node_ids(
+					number_of_identifiers.into(),
+					bucket_node_aggregate_activity.clone(),
+				);
+				let bucket_id = bucket_node_aggregate_activity.clone().bucket_id;
+
+				log::info!("🚀 Merkle Node Identifiers for bucket_node_aggregate_activity: node id: {:?} bucket_id:{:?}  identifiers: {:?}",
+						bucket_node_aggregate_activity.clone().node_id, bucket_id.clone(), merkle_node_ids);
+
+				let challenge_responses = Self::fetch_challenge_responses(
+					cluster_id,
+					bucket_node_aggregate_activity.clone().node_id,
+					era_id,
+					bucket_id,
+					merkle_node_ids,
+					dac_nodes,
+				)
+				.map_err(|err| vec![err])?;
+
+				log::info!("🚀 Fetched challenge response node id: {:?} bucket_id:{:?}  challenge_response: {:?}",
+						bucket_node_aggregate_activity.clone().node_id, bucket_id.clone(), challenge_responses);
+
+				for challenge_response in challenge_responses {
+					for proof in challenge_response.proofs {
+						let leaf_record_hashes: Vec<ActivityHash> =
+							proof.leafs.into_iter().map(|p| p.hash::<T>()).collect();
+
+						let leaf_record_hashes_string: Vec<String> =
+							leaf_record_hashes.clone().into_iter().map(hex::encode).collect();
+
+						log::info!("🚀 Fetched leaf record hashes node id: {:?} bucket_id:{:?}  leaf_record_hashes: {:?}",
+						bucket_node_aggregate_activity.clone().node_id, bucket_id.clone(), leaf_record_hashes_string);
+
+						let leaf_record_root =
```

Comment:
The root of an Aggregate Merkle tree must be calculated by taking Hash from the usage in the Root node (the node with `ID=1` based on the documentation). I don't believe you can calculate it only from tree leaves.

Correct me if I'm wrong:
cc: @khssnv  @upalinski @jwoloch

Reply:
@yahortsaryk no, Hash of root is hash of child hashes.

Reply:
@jwoloch This is based on what @upalinski mentioned earlier
<img width="426" alt="image" src="https://github.com/user-attachments/assets/880d7649-2c68-4db6-a4e0-9903645e7491">